### PR TITLE
Fix bug in manual instrumentation

### PR DESCRIPTION
**Pull request by andrewlock**
## Summary of changes

Fixes the `Tracer.Settings` getter in manual instrumentation

## Reason for change

While investigating something else, I noticed that the `Tracer.Settings` getter doesn't work correctly if the tracer has been reconfigured. The previous implementation assumed that the automatic `Tracer` instance is tied to a given `ImmutableTracerSettings` object, but that's not the case. It's the `TracerManager` that's tied to a settings object, but in "normal" usage, the `Tracer` instance _isn't_ tied to a single `TracerManager`, and hence _isn't_ tied to a settings object. e.g

```csharp
Tracer.Instance.ActiveScope;// Trigger the client-side creation of the `Tracer` object with default settings

// customer manually re-configures tracer
var settings = TracerSettings.FromDefaultSources();
settings.ServiceName = "updated-name";
settings.Environment = "updated-env";
settings.GlobalTags = new Dictionary<string, string> { { "Updated-key", "Updated Value" } };
Tracer.Configure(settings); // update Tracer.Instance

// Fetch the settings back
var immutable = Tracer.Instance.Settings;
Debug.Assert(immutable.ServiceName == settings.ServiceName); // fails
Debug.Assert(immutable.Environment == settings.Environment ); // fails
// The settings have not changed from the _original_ values
```

## Implementation details

Instead of populating a settings object when creating the manual `Tracer` instance, create it when the customer explicitly calls `Tracer.Settings`. This is likely to be rare (will add telemetry in a followup PR) but we also don't want to create the dictionary transport every time they call it. To work around this, we track the "current" (automatic) `ImmutableTracerSettings` in a local field and pass that to the automatic side, so that it can check whether the settings need to be updated or not.

## Test coverage

Added additional assertions to the manual instrumentation tests, which fail prior to these changes

## Other details

This _removes_ an instrumentation point (and adds another) but we keep the instrumentation around for back-compat reasons

### CODE REVIEW

**Comment by andrewlock** [comment]
## Execution-Time Benchmarks Report :stopwatch:

Execution-time results for samples comparing the following branches/commits:
- [This PR (6330)](https://github.com/DataDog/dd-trace-dotnet/tree/ee4d15fb92dd94f79777eef8c56b36a6043caa71)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/7826d354d492243718d33f96469683a73f0c151e)

Execution-time benchmarks measure the whole time it takes to execute a program. And are intended to measure the one-off costs. Cases where the execution time results for the PR are worse than latest master results are shown in **red**. The following thresholds were used for comparing the execution times:
* Welch test with statistical test for significance of **5%**
* Only results indicating a difference greater than **5%** and **5 ms** are considered.


Note that these results are based on a _single_ point-in-time result for each branch. For full results, see the [dashboard](https://ddstaging.datadoghq.com/dashboard/4qn-6fi-54p/apm-net-execution-time-benchmarks).

Graphs show the p99 interval based on the mean and StdDev of the test run, as well as the mean value of the run (shown as a diamond below the graph).
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6330) - mean (71ms)  : 64, 79
     .   : milestone, 71,
    master - mean (71ms)  : 64, 79
     .   : milestone, 71,

    section CallTarget+Inlining+NGEN
    This PR (6330) - mean (1,107ms)  : 1084, 1129
     .   : milestone, 1107,
    master - mean (1,105ms)  : 1089, 1120
     .   : milestone, 1105,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6330) - mean (108ms)  : 105, 111
     .   : milestone, 108,
    master - mean (108ms)  : 106, 110
     .   : milestone, 108,

    section CallTarget+Inlining+NGEN
    This PR (6330) - mean (766ms)  : 752, 780
     .   : milestone, 766,
    master - mean (768ms)  : 748, 787
     .   : milestone, 768,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6330) - mean (92ms)  : 90, 94
     .   : milestone, 92,
    master - mean (92ms)  : 89, 94
     .   : milestone, 92,

    section CallTarget+Inlining+NGEN
    This PR (6330) - mean (718ms)  : 703, 733
     .   : milestone, 718,
    master - mean (722ms)  : 705, 738
     .   : milestone, 722,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6330) - mean (191ms)  : 187, 195
     .   : milestone, 191,
    master - mean (191ms)  : 187, 195
     .   : milestone, 191,

    section CallTarget+Inlining+NGEN
    This PR (6330) - mean (1,214ms)  : 1195, 1233
     .   : milestone, 1214,
    master - mean (1,210ms)  : 1190, 1231
     .   : milestone, 1210,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6330) - mean (277ms)  : 273, 281
     .   : milestone, 277,
    master - mean (276ms)  : 272, 281
     .   : milestone, 276,

    section CallTarget+Inlining+NGEN
    This PR (6330) - mean (946ms)  : 931, 961
     .   : milestone, 946,
    master - mean (947ms)  : 929, 964
     .   : milestone, 947,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6330) - mean (266ms)  : 262, 269
     .   : milestone, 266,
    master - mean (267ms)  : 262, 271
     .   : milestone, 267,

    section CallTarget+Inlining+NGEN
    This PR (6330) - mean (923ms)  : 908, 939
     .   : milestone, 923,
    master - mean (931ms)  : 915, 947
     .   : milestone, 931,

```

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/ManualInstrumentation/Tracer/GetUpdatedImmutableTracerSettingsIntegration.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+/// <summary>
+/// Datadog.Trace.Tracer::GetUpdatedImmutableTracerSettings() calltarget instrumentation
+/// </summary>
+[InstrumentMethod(
+    AssemblyName = "Datadog.Trace.Manual",
+    TypeName = "Datadog.Trace.Tracer",
+    MethodName = "GetUpdatedImmutableTracerSettings",
+    ReturnTypeName = "System.Collections.Generic.Dictionary`2[System.String,System.Object]",
+    ParameterTypeNames = [ClrNames.Object, "System.Object&"],
+    MinimumVersion = "3.6.*", // added in 3.7.0
```
Comment (kevingosse) [inline tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/ManualInstrumentation/Tracer/GetUpdatedImmutableTracerSettingsIntegration.cs:23]: Shouldn't minimum version be 3.7.0 then?

File: tracer/src/Datadog.Trace.Manual/Tracer.cs [hunk 2]
```diff
@@ -116,6 +136,13 @@ private static void Configure(Dictionary<string, object?> settings)
         [Instrumented]
         private static object? GetAutomaticTracerInstance() => null;
 
+        /// <summary>
+        /// Automatic instrumentation intercepts this method and returns a dictionary populated with updated
+        /// settings, only if the ImmutableTracerSettings (automatic) provided is different to the current one.
+        /// </summary>
+        [Instrumented]
+        private Dictionary<string, object?>? GetUpdatedImmutableTracerSettings(object? automaticTracer, ref object? automaticSettings) => null;
```
[the remark this answers is not in the document]
Comment (andrewlock) [reply tracer/src/Datadog.Trace.Manual/Tracer.cs]: Fine, since you clearly feel incredibly strongly about it, I've made the change.

**Review by kevingosse — APPROVED** [review]

File: tracer/src/Datadog.Trace.Manual/Tracer.cs [hunk 3]
```diff
@@ -16,14 +17,29 @@ namespace Datadog.Trace
...
         private Tracer(object? automaticTracer, Dictionary<string, object?> initialValues)
         {
             AutomaticTracer = automaticTracer;
-            Settings = new ImmutableTracerSettings(initialValues);
+
+            // In 3.7.0+ we don't bother populating the dictionary with the settings
+            // seeing as the whole _settings object is going to be thrown away and
+            // repopulated if the customer calls Settings. Kind of annoying...
+            // so to avoid the extra work/allocation we check whether the initial values contains the
+            // agent key, as an easy way of detecting if we're running in this version-conflict mode.
```
Comment (lucaspimentel) [inline tracer/src/Datadog.Trace.Manual/Tracer.cs:33]: Why `AgentUriKey`? Is there something special about that settings? Should you include that in this comment?

**Review by lucaspimentel — APPROVED** [review]

File: tracer/src/Datadog.Trace.Manual/Tracer.cs (hunk 3 above)
Comment (andrewlock) [reply tracer/src/Datadog.Trace.Manual/Tracer.cs:33]: No, there's nothing special about it 😄 It's just one that we already support and probably will _continue_ to support for all future versions

**Review by bouwkast — APPROVED** [review]
Thanks :)

**Pull request merged**